### DISCUSSION TITLE
Firmware binary upstreaming

### DISCUSSION THREAD

**Discussion by lgirdwood** [thesofproject/sof#7378] [Team Posts]
Now that v1.3 is done, I will be upstreaming the firmware binaries.
I would like to use soft links in the FW repos so that we can have multiple versions available i.e.
```sof-apl-ri -> sof-apl-v1.3.1.ri```
Any objections to this ? This allows users to revert if needed.
Also at what point should we remove old versions from the repo ? year old, 4 versions old ??

**Comment by lgirdwood** [comment]
@dbaluta I guess you will be upstreaming i.MX FW binaries (since you will have xcc for your ISA) ?

**Comment by plbossart** [comment]
the symlinks are fine, it's the same as currently done in /lib/firmware/intel.
I don't know if we need to define any rules to remove stuff for now, our problem is more adding new firmwares.
The one thing I still don't get is how we are going to deal with firmware signed with the Intel production keys. We may need a sim link at the directory level as well, e.v. /lib/firmware/intel/sof -> /lib/firmware/intel/sof-signed or use a kernel parameter to override the default location for the firmware (already implemented).

**Comment by lgirdwood** [comment]
@dbaluta one more thing, you probably want to add imx to README.md now. This means it will be prominent on the main SOF repo page.

**Comment by lgirdwood** [comment]
@plbossart good point. We could have several binaries for one platform signed by different keys. The SOF driver could try different locations / binaries if authentication was failing. i.e. we could have
```lib/fimrware/intel/sof/key/intel and lib/firmware/intel/sof/key/open``` and try ```sof-apl.ri``` in them both ?

**Comment by plbossart** [comment]
I don't know if it makes sense to retry, it'll increase the boot time significantly with the existing timeouts. Better try and find a quirk or something to identify which version should be used.

**Comment by lgirdwood** [comment]
Works for me. Default is Intel signed unless DMI name matches table for open key  ?

**Comment by plbossart** [comment]
actually yes, that's pretty easy to do. I already have the quirks for this but I was using it to kick the SST/cAVS driver out. I can actually change the location for those open devices (Up2, Chromebooks, etc). Good suggestion, need to complete this work.
